internal/metrics/waste.go

```diff
@@ -150,33 +229,66 @@ func (r *wasteMetricsReporter) markAndSlowLog(pod *v1.Pod, tag tagInfo, duration
 	metrics.FromContext(r.ctx).Histogram(schedulingWastePerInstanceGroup, tag.tag, InstanceGroupTag(r.ctx, instanceGroup)).Update(duration.Nanoseconds())
 }
 
-func (r *wasteMetricsReporter) onDemandFulfilled(demand *v1alpha1.Demand) {
+func (r *WasteMetricsReporter) onDemandFulfilled(demand *v1alpha1.Demand) {
 	r.lock.Lock()
 	defer r.lock.Unlock()
-	r.info[podKey{demand.Namespace, utils.PodName(demand)}] = demandInfo{
-		demandFulfilledTime: time.Now(),
-		demandCreationTime:  demand.CreationTimestamp.Time,
+
+	podKey := podKey{demand.Namespace, utils.PodName(demand)}
+	info, ok := r.info[podKey]
+	if ok {
+		r.info[podKey] = schedulingMetricInfo{
+			lastFailedSchedulingAttemptInfo: info.lastFailedSchedulingAttemptInfo,
+			demandFulfilledTime:             time.Now(),
+			demandCreationTime:              demand.CreationTimestamp.Time,
+		}
+	} else {
+		r.info[podKey] = schedulingMetricInfo{
+			demandFulfilledTime: time.Now(),
+			demandCreationTime:  demand.CreationTimestamp.Time,
+		}
 	}
 }
 
-func (r *wasteMetricsReporter) onDemandCreated(obj interface{}) {
+func (r *WasteMetricsReporter) onDemandCreated(obj interface{}) {
 	r.lock.Lock()
 	defer r.lock.Unlock()
 	demand, ok := obj.(*v1alpha1.Demand)
 	if !ok {
 		svc1log.FromContext(r.ctx).Error("failed to parse obj as demand")
 		return
 	}
-	r.info[podKey{demand.Namespace, utils.PodName(demand)}] = demandInfo{
-		demandCreationTime: demand.CreationTimestamp.Time,
+	podKey := podKey{demand.Namespace, utils.PodName(demand)}
+	info, ok := r.info[podKey]
```

Comment:
it feels like to prevent all the branching on creating or merging the values in this map, we can change the type we store to `map[podkey]*schedulingMetricInfo`, I believe we can have a helper method to get a value from the map or create an empty struct if absent, and then do:
`r.info[podKey].demandCreationTime = demand.CreationTimestamp.Time`